pkg/gardenlet/controller/shoot/care/health.go

```diff
@@ -736,6 +794,35 @@ func CheckNodeAgentLeases(nodeList *corev1.NodeList, leaseList *coordinationv1.L
 	return nil
 }
 
+// CheckForExpiredNodeLeases checks if the number of expired node Leases surpasses 20% of all existing Leases. If yes,
+// an error will be returned. The motivation is that dependency-watchdog is starting to scale down controllers when 60%
+// of the Leases are expired.
+func CheckForExpiredNodeLeases(nodeList *corev1.NodeList, leaseList *coordinationv1.LeaseList, clock clock.Clock) error {
+	if len(leaseList.Items) == 0 || len(nodeList.Items) == 0 {
+		return nil
+	}
+
+	nodeNames := sets.Set[string]{}
+	for _, node := range nodeList.Items {
+		nodeNames.Insert(node.Name)
+	}
+
+	var expiredLeases int
+	for _, lease := range leaseList.Items {
+		// we only care about Leases related to existing nodes
+		if nodeNames.Has(lease.Name) &&
+			lease.Spec.RenewTime.Add(time.Second*time.Duration(*lease.Spec.LeaseDurationSeconds)).Before(clock.Now()) {
```

Comment:
FYI:- While this is the correct way of calculating the expired leases, DWD may act before the lease actually expires in cases where the `nodeMonitorGracePeriod` of KCM is equal to the `nodeLeaseDurationSeconds ` of the kubelet. In DWD, we calculate the "expiry" of a lease as follows (ref- https://github.com/gardener/dependency-watchdog/blob/55cc3697b3cb3d6547dadbedb8a48ae51283ee56/internal/prober/prober.go#L175) :-
```
revisedNodeLeaseExpiryTime := float64(p.config.KCMNodeMonitorGraceDuration.Duration) * expiryBufferFraction
expiryTime := lease.Spec.RenewTime.Add(time.Duration(revisedNodeLeaseExpiryTime))
``` 

The `expiryBufferFraction` is hardcoded to `0.75` at the moment. We can make this configurable if you want. So as per the above calculation if `nodeMonitorGracePeriod` of KCM and `nodeLeaseDurationSeconds` of kubelet is `40s` (default value for both), DWD will take action before the leases are expired. This is done because we cannot wait for the lease to expire as KCM will also immediately mark the Node as `Unknown` and start evicting pods. We want the DWD to scale down KCM before this happens.

Reply:
I've seen this in the DWD code, but is this level precision really relevant for this "health check"? The idea is to notify operators if there is something wrong in the cluster. If DWD is faster and decides to scale down when >= `60%` are reached, then that's fine. With the first check, even this will now become noticeable by operators.

Reply:
Then can we change this log 
```
 fmt.Errorf("%d%% of all Leases in %s namespace are expired - dependency-watchdog-prober will start scaling down controllers when 60%% is reached", expiredLeasesPercentage, corev1.NamespaceNodeLease)
```

to

```
 fmt.Errorf("%d%% of all Leases in %s namespace are expired ", expiredLeasesPercentage, corev1.NamespaceNodeLease)
```

Because DWD may have already done the scaling.

Reply:
I think it makes sense to note that this is somewhat related to DWD-prober. I've still adapted the message, PTAL.